File: plugins/statusline/check-update.js
#!/usr/bin/env node
// Claude Code Update Checker
// Checks npm registry for newer versions, fetches GitHub release notes,
// and caches results to ~/.claude/statusline/update-cache.json
// Runs from SessionStart hook — outputs notification if update available.

const fs = require('fs');
const path = require('path');
const os = require('os');
const { execSync } = require('child_process');
const https = require('https');

const CACHE_DIR = path.join(os.homedir(), '.claude', 'statusline');
const CACHE_FILE = path.join(CACHE_DIR, 'update-cache.json');
const CHECK_INTERVAL_MS = 4 * 3600 * 1000; // 4 hours

function readCache() {
  try {
    return JSON.parse(fs.readFileSync(CACHE_FILE, 'utf8'));
  } catch {
    return null;
  }
}

function writeCache(data) {
  try {
    fs.mkdirSync(CACHE_DIR, { recursive: true });
    fs.writeFileSync(CACHE_FILE, JSON.stringify(data, null, 2) + '\n');
  } catch {}
}

function getCurrentVersion() {
  // 1. Try claude in PATH (works when check-update.sh sets up PATH correctly)
  const commands = ['claude --version', 'claude-code --version'];
  for (const cmd of commands) {
    try {
      const output = execSync(cmd, {
        encoding: 'utf8', timeout: 5000,
        stdio: ['pipe', 'pipe', 'pipe'],
      }).trim();
      const match = output.match(/(\d+\.\d+\.\d+)/);
      if (match) return match[1];
    } catch {}
  }

  // 2. Try claude binary at common locations and relative to node
  const homedir = os.homedir();
  const candidatePaths = [
    path.join(path.dirname(process.execPath), 'claude'),  // same dir as node
    path.join(homedir, '.npm-global', 'bin', 'claude'),
    path.join(homedir, '.local', 'share', 'nodeenv', 'bin', 'claude'),
    path.join(homedir, '.volta', 'bin', 'claude'),
    '/usr/local/bin/claude',
    '/usr/bin/claude',
    '/opt/homebrew/bin/claude',
  ];
  for (const p of candidatePaths) {
    try {
      if (!fs.existsSync(p)) continue;
      const output = execSync(`"${p}" --version`, {
        encoding: 'utf8', timeout: 5000,
        stdio: ['pipe', 'pipe', 'pipe'],
      }).trim();
      const match = output.match(/(\d+\.\d+\.\d+)/);
      if (match) return match[1];
    } catch {}
  }

  // 3. Try reading version from npm global package.json
  try {
    const globalRoot = execSync('npm root -g', {
      encoding: 'utf8', timeout: 5000,
      stdio: ['pipe', 'pipe', 'pipe'],
    }).trim();
    const pkgPath = path.join(globalRoot, '@anthropic-ai', 'claude-code', 'package.json');
    const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
    if (pkg.version) return pkg.version;
  } catch {}

  return null;
}

function fetchJSON(url) {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url);
    const opts = {
      hostname: parsed.hostname,
      path: parsed.pathname + parsed.search,
      headers: { 'User-Agent': 'claude-code-statusline-plugin/1.0' },
    };
    const req = https.get(opts, res => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return fetchJSON(res.headers.location).then(resolve).catch(reject);
      }
      if (res.statusCode !== 200) {
        return reject(new Error(`HTTP ${res.statusCode}`));
      }
      let body = '';
      res.on('data', chunk => (body += chunk));
      res.on('end', () => {
        try { resolve(JSON.parse(body)); }
        catch (e) { reject(e); }
      });
    });
    req.on('error', reject);
    req.setTimeout(8000, () => { req.destroy(); reject(new Error('timeout')); });
  });
}

function isNewer(latest, current) {
  const l = latest.split('.').map(Number);
  const c = current.split('.').map(Number);
  for (let i = 0; i < 3; i++) {
    if ((l[i] || 0) > (c[i] || 0)) return true;
    if ((l[i] || 0) < (c[i] || 0)) return false;
  }
  return false;
}

function formatNotification(cache) {
  if (!cache || !cache.has_update) return '';
  let msg = `\u2b06 Claude Code v${cache.latest_version} is available! (current: v${cache.current_version})`;
  if (cache.release_notes) {
    msg += `\n\nWhat's new in v${cache.latest_version}:\n${cache.release_notes}`;
  }
  return msg;
}

async function main() {
  try {
    // Check cooldown — re-display cached notification but skip network
    const cache = readCache();
    if (cache && cache.checked_at && (Date.now() - cache.checked_at < CHECK_INTERVAL_MS)) {
      const msg = formatNotification(cache);
      if (msg) process.stdout.write(msg + '\n');
      return;
    }

    const currentVersion = getCurrentVersion();
    if (!currentVersion) {
      writeCache({ checked_at: Date.now(), error: 'cannot determine current version' });
      return;
    }

    // Fetch latest version from npm dist-tags
    let latestVersion;
    try {
      const tags = await fetchJSON(
        'https://registry.npmjs.org/-/package/@anthropic-ai/claude-code/dist-tags'
      );
      latestVersion = tags.latest;
    } catch {
      writeCache({
        checked_at: Date.now(),
        current_version: currentVersion,
        error: 'npm registry unreachable',
      });
      return;
    }

    if (!latestVersion) {
      writeCache({
        checked_at: Date.now(),
        current_version: currentVersion,
        error: 'no latest tag found',
      });
      return;
    }

    const hasUpdate = isNewer(latestVersion, currentVersion);
    let releaseNotes = '';

    if (hasUpdate) {
      // Try GitHub releases API (try both v-prefixed and plain tags)
      for (const tag of [`v${latestVersion}`, latestVersion]) {
        try {
          const release = await fetchJSON(
            `https://api.github.com/repos/anthropics/claude-code/releases/tags/${tag}`
          );
          if (release.body) {
            releaseNotes = release.body;
            break;
          }
        } catch {}
      }

      // Also try "latest" endpoint if tag lookup failed
      if (!releaseNotes) {
        try {
          const release = await fetchJSON(
            'https://api.github.com/repos/anthropics/claude-code/releases/latest'
          );
          if (release.body && release.tag_name) {
            const tagVer = release.tag_name.replace(/^v/, '');
            // Only use if it matches the latest npm version
            if (tagVer === latestVersion) {
              releaseNotes = release.body;
            }
          }
        } catch {}
      }

      // Trim excessively long notes
      if (releaseNotes.length > 1500) {
        releaseNotes = releaseNotes.substring(0, 1500) + '\n…(truncated)';
      }
    }

    const result = {
      checked_at: Date.now(),
      current_version: currentVersion,
      latest_version: latestVersion,
      has_update: hasUpdate,
      release_notes: releaseNotes,
    };

    writeCache(result);

    const msg = formatNotification(result);
    if (msg) process.stdout.write(msg + '\n');
  } catch {
    // Silent fail — never break session start
  }
}

main().then(() => process.exit(0)).catch(() => process.exit(0));

File: plugins/statusline/check-update.py
#!/usr/bin/env python3
"""Claude Code Update Checker — Python fallback (no Node.js required)
Checks npm registry for newer versions, fetches GitHub release notes,
caches results to ~/.claude/statusline/update-cache.json.
Uses urllib (stdlib) for HTTP, falls back to curl."""
import sys, os, json, re, time, subprocess

CACHE_DIR = os.path.join(os.path.expanduser('~'), '.claude', 'statusline')
CACHE_FILE = os.path.join(CACHE_DIR, 'update-cache.json')
CHECK_INTERVAL_MS = 4 * 3600 * 1000  # 4 hours in ms


def read_cache():
    try:
        with open(CACHE_FILE) as f:
            return json.load(f)
    except Exception:
        return None


def write_cache(data):
    try:
        os.makedirs(CACHE_DIR, exist_ok=True)
        with open(CACHE_FILE, 'w') as f:
            json.dump(data, f, indent=2)
            f.write('\n')
    except Exception:
        pass


def get_current_version():
    """Find current Claude Code version via binary or install metadata."""
    home = os.path.expanduser('~')

    # 1. Try claude/claude-code in PATH
    for cmd in [['claude', '--version'], ['claude-code', '--version']]:
        try:
            r = subprocess.run(cmd, capture_output=True, text=True, timeout=5)
            m = re.search(r'(\d+\.\d+\.\d+)', r.stdout)
            if m:
                return m.group(1)
        except Exception:
            pass

    # 2. Try common binary locations
    for p in [
        os.path.join(home, '.local', 'bin', 'claude'),
        os.path.join(home, '.npm-global', 'bin', 'claude'),
        os.path.join(home, '.local', 'share', 'nodeenv', 'bin', 'claude'),
        os.path.join(home, '.volta', 'bin', 'claude'),
        '/usr/local/bin/claude',
        '/usr/bin/claude',
        '/opt/homebrew/bin/claude',
    ]:
        if os.path.isfile(p) and os.access(p, os.X_OK):
            try:
                r = subprocess.run([p, '--version'], capture_output=True, text=True, timeout=5)
                m = re.search(r'(\d+\.\d+\.\d+)', r.stdout)
                if m:
                    return m.group(1)
            except Exception:
                pass

    # 3. Try reading version symlink (native installer layout)
    # ~/.local/bin/claude -> ~/.local/share/claude/versions/X.Y.Z
    link = os.path.join(home, '.local', 'bin', 'claude')
    try:
        target = os.readlink(link)
        m = re.search(r'(\d+\.\d+\.\d+)', target)
        if m:
            return m.group(1)
    except Exception:
        pass

    return None


def fetch_json(url):
    """Fetch JSON from URL using urllib, falling back to curl."""
    # Try urllib first
    try:
        import urllib.request
        req = urllib.request.Request(
            url, headers={'User-Agent': 'claude-code-statusline-plugin/1.0'}
        )
        with urllib.request.urlopen(req, timeout=8) as resp:
            return json.loads(resp.read().decode())
    except Exception:
        pass
    # Fall back to curl
    try:
        r = subprocess.run(
            ['curl', '-sf', '-H', 'User-Agent: claude-code-statusline-plugin/1.0', url],
            capture_output=True, text=True, timeout=10,
        )
        if r.returncode == 0 and r.stdout.strip():
            return json.loads(r.stdout)
    except Exception:
        pass
    return None


def is_newer(latest, current):
    l = [int(x) for x in latest.split('.')]
    c = [int(x) for x in current.split('.')]
    for i in range(3):
        lv = l[i] if i < len(l) else 0
        cv = c[i] if i < len(c) else 0
        if lv > cv:
            return True
        if lv < cv:
            return False
    return False


def format_notification(cache):
    if not cache or not cache.get('has_update'):
        return ''
    msg = f'\u2b06 Claude Code v{cache["latest_version"]} is available! (current: v{cache["current_version"]})'
    notes = cache.get('release_notes', '')
    if notes:
        msg += f'\n\nWhat\'s new in v{cache["latest_version"]}:\n{notes}'
    return msg


def main():
    try:
        now = time.time() * 1000  # ms to match JS cache format

        # Check cooldown
        cache = read_cache()
        if cache and cache.get('checked_at') and (now - cache['checked_at'] < CHECK_INTERVAL_MS):
            msg = format_notification(cache)
            if msg:
                sys.stdout.write(msg + '\n')
            return

        current = get_current_version()
        if not current:
            write_cache({'checked_at': now, 'error': 'cannot determine current version'})
            return

        # Fetch latest from npm dist-tags
        tags = fetch_json(
            'https://registry.npmjs.org/-/package/@anthropic-ai/claude-code/dist-tags'
        )
        if not tags or not tags.get('latest'):
            write_cache({
                'checked_at': now,
                'current_version': current,
                'error': 'npm fetch failed',
            })
            return

        latest = tags['latest']
        has_update = is_newer(latest, current)
        release_notes = ''

        if has_update:
            # Try GitHub releases (v-prefixed and plain tags)
            for tag in [f'v{latest}', latest]:
                data = fetch_json(
                    f'https://api.github.com/repos/anthropics/claude-code/releases/tags/{tag}'
                )
                if data and data.get('body'):
                    release_notes = data['body']
                    break

            # Try "latest" endpoint as last resort
            if not release_notes:
                data = fetch_json(
                    'https://api.github.com/repos/anthropics/claude-code/releases/latest'
                )
                if (data and data.get('body')
                        and data.get('tag_name', '').lstrip('v') == latest):
                    release_notes = data['body']

            if len(release_notes) > 1500:
                release_notes = release_notes[:1500] + '\n\u2026(truncated)'

        result = {
            'checked_at': now,
            'current_version': current,
            'latest_version': latest,
            'has_update': has_update,
            'release_notes': release_notes,
        }
        write_cache(result)

        msg = format_notification(result)
        if msg:
            sys.stdout.write(msg + '\n')
    except Exception:
        pass


if __name__ == '__main__':
    main()
